\documentclass[9pt]{beamer}
\usetheme{cmepda}

\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}

\graphicspath{{figures/}} 


\title{Advanced python features}
\subtitle{Computing Methods for Experimental Physics and Data Analysis}
\date{Compiled on \today}
\author{A. Manfreda}
\institute[INFN]{INFN--Pisa}
\email{alberto.manfreda@pi.infn.it}


\begin{document}


\titleframe


\begin{frame}
  \frametitle{Errors and Exceptions}
  
  \begin{itemize}
    \item \alert{Error handling} is one of the most important problem to solve when designing a program
    \item What should I do when I piece of code fails?
    \item What does fail mean?
    \begin {itemize}
      \item Invalid input e.g. passing a path to a non existent file, or passing a string to a function for dividing numbers
      \item Valid output not found, e.g searching the position of the letter 'd' in the string 'elephant'
      \item Output cannot be find in a reasonable amount of time
      \item Runtime resource failures: network connection down, disk space ended\dots
    \end{itemize}
    \item Two phylosophies (historically):
    \begin{itemize}
      \item Return some \alert{error flag} (in different ways) to tell the user that something went wrong
      \item \alert{Exceptions}
    \end{itemize}
    \item Example: a typical convention for programs is to return 0 from the main if the execution was successful and an
          error code (integer number) otherwise
  \end{itemize}
  
\end{frame}


\begin{frame}
  \frametitle{Error flags}
  \input{pygments/error_flags}
\end{frame}


\begin{frame}
  \frametitle{Problems of error flags}
  
    Error codes have their use (and are fine in some cases) but they suffer from a few issues:
    \begin{itemize}
      \item Choosing them is often arbitrary (and sometimes is difficult to make a sensible choice)
      \begin{itemize}
        \item What if all the numbers can represent meaningful output of the function?
      \end{itemize}
      \item Are cumbersome to use
      \begin{itemize}
        \item Which error flag is used by a function? 0? -1? 99999999? $\rightarrow$ you have to go through the documentation for each!
        \item If you have a deep hierarchy of functions you have to perform checks and pass the error up at every level!
      \end{itemize}
      \item What if the caller of a function does not check the error flag?
      \begin{itemize}
        \item The bug can propagate \alert{silently} through its code!
      \end{itemize}
    \end{itemize}
   
    \medskip
   
    We want something that:
    \begin{itemize}
      \item Is clearly separated from the returned output
      \item Cannot be silently ignored by the user
      \item Is easy to report to upper level without lots of lines of code
    \end{itemize}
  
\end{frame}


\begin{frame}
  \frametitle{A different way}
  \input{pygments/exceptions_vs_err_flags}
\end{frame}



\begin{frame}
  \frametitle{Enter exceptions}

  \begin{itemize}
    \item An exception is an object that can be \alert{raised} (in other languages also \textit{thrown}) by
          a piece of code to signal that something went wrong
    \item When an exception is raised the normal flow of the code is interrupted
    \item The program automatically propagate the exception back in the function hierarchy
          until it found a place where the exception is  \alert{catched} and handled
    \item If the exception is never catched, not even in the main, the program crash \alert{with a specific error message}
    \item Cathcing the exception is done with a \emph{try - except} block
  \end{itemize}
\end{frame}


\begin{frame}
  \frametitle{Exceptions}
  \input{pygments/exceptions}
\end{frame}


\begin{frame}
  \frametitle{The beauty of throwning stuff}

  \begin{itemize}
    \item If that was all, exceptions would only be moderately useful
    \item The real bargain is that you can send back information together with the exception
    \item In fact you \textit{are sending a full object}: the excetpion iteslf. Surprised?
    \item Inside the exception you can report all kind of data useful to reconstruct the exact error,
          which can be used by the caller for debug or to produce meaningful error messages
    \item You can also select which exceptions you catch, leaving the others propagate up
    \item Python provides a rich hierarchy of exception classes, which you can further customize
          (if you want) by deriving your own subclasses
  \end{itemize}
\end{frame}


\begin{frame}
  \frametitle{The family tree of Python exceptions}
  \centering\includegraphics[height=0.8\textheight]{python_exceptions}
\end{frame}


\begin{frame}
  \frametitle{Catching exceptions}
  \input{pygments/try_block}
\end{frame}


\begin{frame}
  \frametitle{Raising exceptions}
  \input{pygments/raising}
\end{frame}


\begin{frame}
  \frametitle{Where to catch exceptions?}

  \begin{itemize}
    \item Differently from error flags, which needs to be checked as early as
          possible, you are not in a rush with exceptions
    \item Remeber: your goal is to provide the user a meaningful error message and
          useful debug information.
    \item You should catch an exception only when you have enough context to
          do that - which sometimes means waiting a few levels in the hierarchy!
  \end{itemize}
\end{frame}


\begin{frame}
  \frametitle{When to catch?}
  \centering
  \includegraphics[width=0.8\textwidth]{lab_file.png}
\end{frame}



\begin{frame}
  \frametitle{When to catch?}
  \input{pygments/when_to_catch}
\end{frame}


\begin{frame}
  \frametitle{Catch too early}
  \input{pygments/when_to_catch_1}
\end{frame}


\begin{frame}
  \frametitle{Catch when needed}
  \input{pygments/when_to_catch_2}
\end{frame}

\begin{frame}
  \frametitle{There is no check - only try}

  \begin{itemize}
    \item In Python exceptions are the default methods for handling failures
    \item Many functions raise an exception when something goes wrong
    \item The common approach is: do not chech the input beforehand. Use it and
          be ready to catch excpetions if any.
    \item \textit{Easier to ask for forgiveness than permission.} 
  \end{itemize}
\end{frame}


\begin{frame}
  \frametitle{Easier to ask for forgiveness}
  \input{pygments/dont_ask_permission}
\end{frame}


\begin{frame}
  \frametitle{}
    \centering \Large Decorators
\end{frame}


\begin{frame}
  \frametitle{Functions inside functions}
  \begin{itemize}
    \item Function in python are \alert{first class object}
    \item The name is a bit misleading, but what actually means is that functions
          can be passed as argument to other functions and returned as result from
          other functions
    \item This shouldn't surprise you much: functions are objects of a
          '\emph{function}' class, so they behave like any other vairable in
          Python
    \item Another thing you can (and sometimes want to) do is \emph{defining} a
          function inside another.
    \item Let's see how it works
  \end{itemize}
  
\end{frame}


\begin{frame}
  \frametitle{Functions inside functions}
  \input{pygments/inner_outer}
\end{frame}


\begin{frame}
  \frametitle{Colsures and free variables}
  \begin{itemize}
    \item When a function is created inside another function it has access to the local variables
          of the outer function, even after its scope ended
    \item This is techincally possible because those varibales are kept in a special
          space of memory, the \alert{closure} of the inner function
    \item Such variables are called \alert{free variables}
    \item Note: if you \emph{assign} to a free variable in the inner function, 
          by default a new, local variable is created instead!
    \item To avoid this you have to explictly declare that you want to access the variable in
          the closure using the \alert{nonlocal} keyword
    \item Remember: \emph{'Explicit is better then implicit'}
  \end{itemize}
  
\end{frame}


\begin{frame}
  \frametitle{Free variables - a mistake to avoid}
  \input{pygments/closure_wrong}
\end{frame}


\begin{frame}
  \frametitle{Free variables - the correct way}
  \input{pygments/closure_right}
\end{frame}


\begin{frame}
  \frametitle{Wrapping functions}
  \begin{itemize}
    \item The typical use of defining a function inside a function is to create
          a \alert{wrapper}
    \item A wrapper is a function that calls another one adding a layer
          of functionalities in between - for example it may do some pre-process
          of the input, or change the output in some way, or measure the 
          execution time or whatever we want
    \item The techinque for creating a wrapper fucntion in Python is:
    \begin{itemize}
       \item Pass the function that we want to wrap as argument of the outer
             function
       \item Inside the outer function we define an inner function, which is the
             actual wrapper
       \item The wrapper calls the wrapped function and adds its functionalities,
             before and/or after the call. It may return the same output or a 
             manipulated one.
       \item Then from the outer fucntion we return the wrapper
    \end{itemize}
  \end{itemize}
  
\end{frame}


\begin{frame}
  \frametitle{Wrapper}
  \input{pygments/wrapper}
\end{frame}


\begin{frame}
  \frametitle{Decorators}
  \begin{itemize}
    \item Often, when you wrap a function, you don't want to change
          it's name, so you reassign the wrapped funtion to its old name
    \item In fact, this techinque is so common that python introduced a special
          syntax for it: decorators
    \item A decorated function has simply the name of the wrapper added with
          a '@' on top of its declaration
  \end{itemize}
  
\end{frame}


\begin{frame}
  \frametitle{Decorators}
  \input{pygments/decorator}
\end{frame}


\begin{frame}
  \frametitle{A decorator to measure execution time}
  \input{pygments/time_measuring_decor}
\end{frame}


\begin{frame}
  \frametitle{The @classmethod decorator}
  \begin{itemize}
    \item We have already seen a built-in Python decorator: \emph{@property}
    \item We used that to get proper encapsulation
    \item There is another built-in decorator one which is very useful for classes: \alert{\emph{@classmethod}}
    \item A classmethod is like a class attribute: you don't need an instance to
          use it
    \item A class method can access class attributes but not instance attributes
    \item The main use for class methods is to provide \alert{alternate constructors}
  \end{itemize}
  
\end{frame}


\begin{frame}
  \frametitle{Class method}
  \input{pygments/classmethod}
\end{frame}


\begin{frame}
  \frametitle{}
  \centering \Large Generators
\end{frame}



\begin{frame}
  \frametitle{Generators}
  
  \begin{itemize}
    \item We know how to iterate over python built-in container or even 
          create our custom iterables
    \item However that assumes a containers exists $\rightarrow$ memory usage
    \item \alert{Generators} allow you to loop over sequences of items even when
          they don't exist before - they items are just created \alert{lazily} the
          moment they are required
    \item For example you can write a generator to loops over the Fibonacci
          succession. You can't create the sequence earlier, since it is not
          finite!
    \item Generators are created through either \alert{generator expressions} or
          \alert{generator functions}
    \item In real life most of the time you will simply use pre-made functions
          that return a generator, like \emph{range()} (in Python 3)
    \item Generators can be used whenever an iterable is expected
  \end{itemize}
  
\end{frame}


\begin{frame}
  \frametitle{Generators first look}
  \input{pygments/generators}
\end{frame}


\begin{frame}
  \frametitle{Generator functions}
  
  \begin{itemize}
    \item A \alert{generator function} is a function that contains the keyword \alert{yield} at
          least once in his body
    \item When you call a generator function the code is not executed - instead
          a generator object is created and returned (even if you don't have a return statement)
    \item Each call to \emph{next()} on the returned generator will make the function code 
          run until it finds a yield statement
    \item Then the execution is paused and the value of the expression on the right 
          of yiel is returned (yielded) to the caller
    \item A further call of next will resume the execution from where it was suspended
          until the next yield and so on
    \item Eventually, when the function body ends, \emph{StopIteration} is raised
    \item Usually generators functions contain a loop - but it's not mandatory!
  \end{itemize}
  
\end{frame}


\begin{frame}
  \frametitle{Generator functions}
  \input{pygments/generator_functions}
\end{frame}


\begin{frame}
  \frametitle{Infinite sequence generators}
  \input{pygments/fibonacci}
\end{frame}


\begin{frame}
  \frametitle{Anonymous (lambda) functions}
  \begin{itemize}
    \item \alert{Anonymous functions}, or \alert{lambda functions} are a construct typical of \alert{functional programming}
    \item \url{https://en.wikipedia.org/wiki/Lambda_calculus}
    \item \url{https://en.wikipedia.org/wiki/Functional_programming}
    \item In Python a lambda function is essentially a special sintax for creating a function
          on the fly, without giving it a name
    \item They are limited to \alert{a single expression}, which is returned to the user
    \item Many of the typical uses for lambdas are already covered in python by generator expressions and comprehension,
          so this is more like a niche feature of the language
  \end{itemize}
  
\end{frame}


\begin{frame}
  \frametitle{Lambda functions}
  \input{pygments/lambda}
\end{frame}


\end{document}
